### PR TITLE
INFRA-417 Improve driver DSL test stability

### PR DESCRIPTION
**Pull request by rnicoll**
This improves the driver DSL test stability by having the tests wait for the nodes to start before connecting to the node. This is likely to be further improvements we can make, but this improves the state for a relatively small set of changes.

### CODE REVIEW

File: testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt [hunk 1]
```diff
@@ -247,7 +248,7 @@ data class DriverParameters(
         val isDebug: Boolean = false,
         val driverDirectory: Path = Paths.get("build") / "node-driver" / getTimestampAsDirectoryName(),
         val portAllocation: PortAllocation = incrementalPortAllocation(),
-        val debugPortAllocation: PortAllocation = incrementalPortAllocation(),
+        val debugPortAllocation: PortAllocation = portAllocation, // Share port allocator by default
```
Comment (adagys) [inline testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt]: Technically this change isn't needed as `incrementalPortAllocation()` will always return the same allocator instance. I raised a related bug, this port allocation config should be cleaned up as now we only really have one allocation strategy and there's nothing to configure.

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt [hunk 2]
```diff
@@ -218,16 +219,24 @@ class DriverDSLImpl(
         }
     }
 
-    private fun establishRpc(config: NodeConfig, processDeathFuture: CordaFuture<out Process>): CordaFuture<CordaRPCOps> {
+    /**
+     * @param connectRetries number of times to try reconnecting before giving up.
```
Comment (adagys) [inline testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt]: missing parameter?

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt [hunk 3]
```diff
@@ -673,6 +682,7 @@ class DriverDSLImpl(
                     }
             )
             val nodeFuture: CordaFuture<NodeHandle> = nodeAndThreadFuture.flatMap { (node, thread) ->
+                node.node.nodeReadyFuture.get() // Wait for the node to be ready before we connect to the node
```
Comment (adagys) [inline testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt:685]: Hm, I guess this changes semantics a bit, this method is expected to return a future of a starting node. But if there are no immediate issues with other tests, probably ok.

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt [hunk 4]
```diff
@@ -779,6 +789,7 @@ class DriverDSLImpl(
     }
 
     companion object {
+        private val DEFAULT_CONNECT_POLL_INTERVAL: Duration = 100.millis
```
Comment (adagys) [inline testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt]: why is this needed? there is an interval default in the poll method already

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt [hunk 5]
```diff
@@ -218,16 +219,24 @@ class DriverDSLImpl(
         }
     }
 
-    private fun establishRpc(config: NodeConfig, processDeathFuture: CordaFuture<out Process>): CordaFuture<CordaRPCOps> {
+    /**
+     * @param connectRetries number of times to try reconnecting before giving up.
+     */
+    private fun establishRpc(config: NodeConfig,
+                             processDeathFuture: CordaFuture<out Process>,
+                             pollInterval: Duration = DEFAULT_CONNECT_POLL_INTERVAL): CordaFuture<CordaRPCOps> {
```
Comment (adagys) [inline testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt]: the new parameter isn't used?

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt (hunk 2 above)
Comment (rnicoll) [reply testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt]: An experiment I removed but apparently missed that the docs remained. Will fix.

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt (hunk 5 above)
Comment (rnicoll) [reply testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt]: `pollInterval = pollInterval) {` below

File: testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt (hunk 1 above)
Comment (rnicoll) [reply testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt]: Removed `debugPortAllocation`, figure broader cuts to this should wait for later.

Comment (adagys) [reply testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt]: Can we do this though? It's an API break. Maybe best to leave it for when the general cleanup happens (if ever)

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt (hunk 5 above)
Comment (adagys) [reply testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt]: it doesn't seem to be set to anything other than the default value though, and there is already a default value set within the `poll` method

Comment (rnicoll) [reply testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt]: The idea is that the calling code _can_ decide this, and understands that there is a value which defines poll interval, rather than it being defined elsewhere and opaque to the calling code.

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt (hunk 4 above)
Comment (rnicoll) [reply testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt]: Reducing the poll interval to try encouraging the tests to pass a bit faster. The previous value was 500ms.

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt (hunk 3 above)
Comment (rnicoll) [reply testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt:685]: Previously it still didn't return until there was a client connection, so it shouldn't change semantics.

In an ideal world, yes it would return immediately, and the connection would happen when separately (potentially as a process triggered from the node ready future), but we'd need to make wider changes than I want to tackle today.

File: testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt (hunk 1 above)
Comment (rnicoll) [reply testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt]: Getting approval from Matthew & Stefano, we'll see how it goes!

**Review by roastario — DISMISSED** [review]
in _THEORY_ this shouldn't have an impact as all port allocators are now actually just proxies to the global shared-memory one.

**Review by mnesbit — APPROVED** [review]
OK

**Review by adagys — APPROVED** [review]
